### ISSUE TITLE
JIT: devirtualization can't rely on CORINFO_FLG_FINAL for methods

### ISSUE THREAD

**Issue by AndyAyersMS** [dotnet/coreclr#19222]
The presence of an explicit method override (see ECMA-355 II.10.3.2) allows a derived class to override an inherited final method. 

Consider the example below.

Currently the jit sees the declared type for local 0 in `main` is C, and the method in the A::f slot of C's method table is C::f and is marked final. So the jit devirtualizes the call, thinking that any derived class of C must likewise have C::f in that slot, as a final method can't be overridden by inheritance.

However D has installed D::f in the corresponding slot in its method table via an explicit override, and so the callvirt in main should target D::f, and the devirtualization is incorrect.

This means that `final` attributions on methods in some class's method table don't convey any guarantees as to the contents of derived class method tables. Instead the jit can only devirtualize here if it knows the exact type of the object.

This kind of override doesn't seem to be possible in C#, but can be described in IL (as below) or C++/CLI.

cc @dotnet/jit-contrib 
cc @adityamandaleeka who brought this example to my attention.

The fix for this is simple. Am currently starting to look at the impact on CQ.

```
.assembly extern mscorlib { }

.assembly X { }

.method assembly static int32 modopt([mscorlib]System.Runtime.CompilerServices.CallConvCdecl) 
        main() cil managed
{
  .entrypoint
  .vtentry 1 : 1
  // Code size       226 (0xe2)
  .maxstack  2
  .locals ([0] class C c,
           [1] uint8 V_1,
           [2] uint8 V_2,
           [3] uint8 V_3,
           [4] uint8 V_4,
           [5] uint8 V_5,
           [6] class B b,
           [7] uint8 V_7,
           [8] uint8 V_8,
           [9] uint8 V_9,
           [10] class C foo)
  IL_0000:  newobj     instance void C::.ctor()
  IL_0005:  stloc.0

  IL_0091:  newobj     instance void D::.ctor()
  IL_0096:  stloc.0
  IL_0097:  ldloc.0
  IL_0098:  callvirt   instance int32 modopt([mscorlib]System.Runtime.CompilerServices.IsLong) A::f()
  ret
}

.class public auto ansi beforefieldinit A
       extends [mscorlib]System.Object
{
  .method public hidebysig newslot virtual 
          instance int32 modopt([mscorlib]System.Runtime.CompilerServices.IsLong) 
          f() cil managed
  {
    // Code size       2 (0x2)
    .maxstack  1
    IL_0000:  ldc.i4.0
    IL_0001:  ret
  } // end of method A::f

  .method public hidebysig specialname rtspecialname 
          instance void  .ctor() cil managed
  {
    // Code size       7 (0x7)
    .maxstack  1
    IL_0000:  ldarg.0
    IL_0001:  call       instance void [mscorlib]System.Object::.ctor()
    IL_0006:  ret
  } // end of method A::.ctor

} // end of class A

.class public auto ansi beforefieldinit B
       extends A
{
  .method public hidebysig virtual instance int32 modopt([mscorlib]System.Runtime.CompilerServices.IsLong) 
          f() cil managed
  {
    // Code size       2 (0x2)
    .maxstack  1
    IL_0000:  ldc.i4.1
    IL_0001:  ret
  } // end of method B::f

  .method public hidebysig specialname rtspecialname 
          instance void  .ctor() cil managed
  {
    // Code size       7 (0x7)
    .maxstack  1
    IL_0000:  ldarg.0
    IL_0001:  call       instance void [mscorlib]System.Object::.ctor()
    IL_0006:  ret
  } // end of method B::.ctor

} // end of class B

.class public auto ansi beforefieldinit C
       extends B
{
  .method public hidebysig newslot virtual final 
          instance int32 modopt([mscorlib]System.Runtime.CompilerServices.IsLong) 
          f() cil managed
  {
    .override A::f
    // Code size       2 (0x2)
    .maxstack  1
    IL_0000:  ldc.i4.2
    IL_0001:  ret
  } // end of method C::f

  .method public hidebysig specialname rtspecialname 
          instance void  .ctor() cil managed
  {
    // Code size       7 (0x7)
    .maxstack  1
    IL_0000:  ldarg.0
    IL_0001:  call       instance void [mscorlib]System.Object::.ctor()
    IL_0006:  ret
  } // end of method C::.ctor

} // end of class C

.class public auto ansi beforefieldinit D
       extends C
{
  .method public hidebysig newslot virtual final 
          instance int32 modopt([mscorlib]System.Runtime.CompilerServices.IsLong) 
          f() cil managed
  {
    .override A::f
    // Code size       2 (0x2)
    .maxstack  1
    IL_0000:  ldc.i4.3
    IL_0001:  ret
  } // end of method D::f

  .method public hidebysig specialname rtspecialname 
          instance void  .ctor() cil managed
  {
    // Code size       7 (0x7)
    .maxstack  1
    IL_0000:  ldarg.0
    IL_0001:  call       instance void [mscorlib]System.Object::.ctor()
    IL_0006:  ret
  } // end of method D::.ctor

} // end of class D
```

**Comment by adityamandaleeka** [comment]
Thanks for opening this @AndyAyersMS!

**Comment by AndyAyersMS** [comment]
Some idea of the impact of the fix.

A number of methods on RuntimeTypeHandle no longer devirtualize (`IsCollectible`, ...)
Likewise some encoder methods (`GetChars`, ...)
```
Total bytes of diff: 1462 (0.00% of base)
    diff is a regression.

Top file regressions by size (bytes):
        1158 : Microsoft.CodeAnalysis.CSharp.dasm (0.03% of base)
         280 : System.Private.CoreLib.dasm (0.01% of base)
         172 : System.Linq.Expressions.dasm (0.02% of base)
          70 : System.Security.Cryptography.X509Certificates.dasm (0.06% of base)
          39 : System.Private.DataContractSerialization.dasm (0.01% of base)

Top file improvements by size (bytes):
        -162 : System.IO.FileSystem.AccessControl.dasm (-1.01% of base)
        -148 : Microsoft.CodeAnalysis.dasm (-0.01% of base)
         -39 : System.Security.AccessControl.dasm (-0.05% of base)

13 total files with size differences (3 improved, 10 regressed), 116 unchanged.

Top method regressions by size (bytes):
         181 : Microsoft.CodeAnalysis.CSharp.dasm - Microsoft.CodeAnalysis.CSharp.Symbols.SourceEventSymbol:CheckModifiersAndType(ref):this
         121 : Microsoft.CodeAnalysis.CSharp.dasm - Microsoft.CodeAnalysis.CSharp.Symbols.ParameterHelpers:ReportParameterErrors(ref,ref,ref,int,ref)
         101 : Microsoft.CodeAnalysis.CSharp.dasm - Microsoft.CodeAnalysis.CSharp.Symbols.SourceUserDefinedOperatorSymbolBase:CheckUserDefinedConversionSignature(ref):this
          82 : Microsoft.CodeAnalysis.CSharp.dasm - Microsoft.CodeAnalysis.CSharp.Symbols.SourceMemberFieldSymbol:TypeChecks(ref,ref,ref,ref):this
          79 : Microsoft.CodeAnalysis.CSharp.dasm - Microsoft.CodeAnalysis.CSharp.Symbols.SourceMemberFieldSymbol:GetFieldType(ref):ref:this

Top method improvements by size (bytes):
        -162 : System.IO.FileSystem.AccessControl.dasm - System.Security.AccessControl.DirectoryObjectSecurity:GetRules(bool,bool,bool,ref):ref:this
        -101 : Microsoft.CodeAnalysis.CSharp.dasm - Microsoft.CodeAnalysis.CSharp.LocalRewriter:MakeConversion(ref,ref,struct,ref,bool,bool,ref):ref:this
         -88 : Microsoft.CodeAnalysis.CSharp.dasm - Microsoft.CodeAnalysis.CSharp.Binder:GetNearestOtherSymbol(ref,ref):ref
         -80 : System.Security.AccessControl.dasm - System.Security.AccessControl.CommonObjectSecurity:GetRules(bool,bool,bool,ref):ref:this
         -78 : Microsoft.CodeAnalysis.CSharp.dasm - Microsoft.CodeAnalysis.CSharp.Symbols.AbstractTypeParameterMap:GetDebuggerDisplay():ref:this

675 total methods with size differences (251 improved, 424 regressed), 221011 unchanged.
```

**Comment by svick** [comment]
I understand that it's possible to express this in IL, but is it actually valid code according to ECMA-335? Its [§I.8.10.2 Method inheritance](https://www.ecma-international.org/publications/files/ECMA-ST/ECMA-335.pdf#%5B%7B%22num%22%3A1064%2C%22gen%22%3A0%7D%2C%7B%22name%22%3A%22XYZ%22%7D%2C87%2C180%2C0%5D) says:

> Virtual methods can be marked as **final**, in which case, they shall not be overridden in a derived 
object type.

Wouldn't this forbid code like the above? So, could the CLR be changed to reject such code, which means it would not cause issues with devirtualization?

Doing that could break some existing working programs, but I think that should be fine on .Net Core (if anyone does use this, they can stay on an old version of .Net Core for some time).

**Comment by AndyAyersMS** [comment]
Yeah, there is similar language elsewhere. The spec doesn't explicitly address the intersection of override and final except in one spot where the scope seems limited. We perhaps could make a case for it being invalid.

* I.8.5.3.2: To prevent overriding a virtual method use `final` (see §I.8.10.2) rather than relying on limited accessibility
* I.8.11.1:  Virtual method can be marked as `final`, indicating that derived object types are not allowed to override the method implementation. 
* II.15.4.2: This virtual method cannot be overridden by derived classes.
* II.15.4.2.2: `final` methods shall not be overridden by derived classes of this type
* II.22.27: The method indexed by MethodDeclaration shall not be `final` (its Flags.Final shall be 0). 

Seemingly this last rule says at if `A::f` was final then the overrides in `C` and `D` would be invalid. But in our example `A::f` is not final.

C++/CLI spec (ECMA-372) also could be interpreted to say that this is override pattern is invalid:

> 19.4.2 Sealed function modifier ... An implicit, explicit, or (in a CLI class type, a) named override can succeed as long as there is a non-sealed virtual function in at least one of the bases. [Example: Consider the case in which A::f is sealed, but B::f is not. If C inherits from A and B, and tries to implement f, it will succeed, but will only override B::f. end example] 

Here "at least one of the bases" could suggest that since `f` is not fixed in `A` then the override is ok, or could also be read as since `f` is fixed in `C`, the override in `D` is invalid.

@CarolEidt @jkotas any thoughts?

**Comment by jkotas** [comment]
@davidwrighton is the domain expert on virtual override resolution. David, could you please comment on this?

**Comment by CarolEidt** [comment]
Personally, I don't think that this kind of overload *should* be allowed - that is, since D derives from C, it can't override a method that has been declared as final in C.

> The presence of an explicit method override (see ECMA-355 II.10.3.2) allows a derived class to override an inherited final method.

It's not entirely clear to me that that's what it's saying. IMO it's unfortunate that it says "The definition can be marked newslot to always create a new virtual method ...". However, it's not clear to me that the "always" means "even if it's declared final in a parent class".

**Comment by svick** [comment]
@CarolEidt I'm confused, how is `newslot` relevant here?

Since `newslot` effectively creates an unrelated method with the same name (i.e. not an override), I think it should be allowed even when the base class has a `final` method of the same name. That also means a `newslot` method in derived class does not affect devirtualization of a `final` method in the base class.

As I understand it, the issue here is that non-`newslot` methods are allowed when the base class method is `final`.

**Comment by davidwrighton** [comment]
@CarolEidt @svick I agree with svick here. However, I think there is something of a spec hole. We really have a situation where we have 3 virtual slots

A.M
C.M
D.M

The real question is what does the sealed flag do? Does it: Mark every slot which the virtual method implements as sealed (approach 1), or does it mark the slot it is directly associated with as sealed (approach 2). 

Under approach 1 for type C, the vtable looks like
A.M -> C.M and sealed
C.M -> C.M and sealed

Then we have type D, where D.M is newslot, and is marked as overriding A.M, thus leading to the a TypeLoadException.

Alternatively under approach 2 for type C the vtable looks like
A.M -> C.M
C.M -> C.M and sealed

And then D's vtable looks like
A.M -> D.M
C.M -> C.M and sealed
D.M -> D.M and sealed

I am inclined to believe that the correct approach here is approach 2 as it align to common use of the .override keyword to handle interface methods. Imagine that instead of A.M, we were overriding I.M with c# that looks like

interface I {  void M(); }
class A : I
{    void I.M() {};    }

class B : A, I
{    void I.M() { /* Implements I.M in B */    }

In this model we end up with a vtable and interface map that looks like:

A vtable
I.M -> B.I.M
A.I.M -> A.I.M and is sealed

B vtable

I.M -> B.I.M
A.I.M -> A.I.M and is sealed
B.I.M -> B.I.M and is sealed

Thus allowing B.I.M to override the behavior of I.M within B even though it was previously implemented via an sealed method, which looks very much like the behavior of approach 2 referenced above.

**Comment by CarolEidt** [comment]
@svick - I'm the one that's confused. I looked at the wrong section when checking into the reference @AndyAyersMS mentioned above.

So ... ignore my comment above.

I agree with @davidwrighton above. 

However, it would seem that the JIT doesn't need to know the exact type of the object, but rather the type & method on which the call is being made. If it's calling C.f and C.f is final, then that's what it should call. It's only the fact that C.f was a newslot that means that it doesn't make the A.f slot final, just the newslot that it created (C.f). Does that make sense or am I still missing something?

**Comment by AndyAyersMS** [comment]
Option 2 is more or less what the runtime does today.

The discussion here is really about whether knowing that if the method table for type X has some final method F in slot Y, then the jit can safely infer that any type derived from X must also have that same final method F in slot Y of its method table.

Under option 2 that is not a safe inference. 

C's slot for A::f has a final method C::f. 

D derives from C and D's slot for A::f has D::f.

Newslot here is a bit of a red herring -- we could easily construct examples like this without newslot by just choosing different names for the methods in the derived classes.

But then the example wouldn't highlight the difference in override behavior between the implicit case (final blocks overriding) and the explicit case (final generally not relevant).

I find option 2 for virtual methods a bit odd -- it means `final` for virtual methods has a very narrow and limiting impact, and the meaning varies depending on whether the class is introducing the method.

Option 1 seems consistent with most modern language semantics and easily explainable. A virtual final method can't be overriden.

**Comment by CarolEidt** [comment]
> Newslot here is a bit of a red herring -- we could easily construct examples like this without newslot by just choosing different names for the methods in the derived classes.

Maybe it's been too long since I've done any spec work, but I thought that newslot was precisely the issue here. That is, the question is whether the declaration of C::f is making just the newly introduced slot final, or the slot that it overrides as well.

**Comment by AndyAyersMS** [comment]
D can install D::xyz into the A::f slot even if C has done a normal override of A::f with a final method C::f. None of that requires newslot.

Maybe we are not using the same terminology -- I am using "slot" to mean an index into a vtable. So "making a slot final" doesn't parse for me. 

A method in a slot in a particular vtable can be final / sealed.

Option 2 would be fine for the jit if 

A.M -> D.M
C.M -> C.M and sealed
D.M -> D.M and sealed

could be described, but currently final/sealed is a property of a method, not of a method + slot. So today either D.M is sealed or it isn't... it can't be both.

**Comment by AndyAyersMS** [comment]
Ok, thanks... expressing this extra check for devirtualization in the current jit/jitinterface path requires a bit of refactoring, but is probably not too bad.

**Comment by AndyAyersMS** [comment]
@CarolEidt my apologies, you were right -- it appears newslot is indeed required.

If by normal overriding a final method is installed in a slot, then explicit overrides in derived classes can't further update the slot. The runtime checks for this and fails class loads.

So the odd case here is when no normal final override happens, there is an explicit final override, and then a further derived class explictly overrides yet again.

Validating that the derived method slot matches the base method slot distinguishes this case from the normal cases. I've updated the fix in #19228 accordingly and we don't lose any of our current devirtualizations.

Thanks to you and @davidwrighton for helping sort this out.
